Firestore/core/src/firebase/firestore/remote/write_stream.h

```diff
@@ -0,0 +1,117 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_WRITE_STREAM_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_WRITE_STREAM_H_
+
+#if !defined(__OBJC__)
+#error "This header only supports Objective-C++"
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+#include <memory>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/remote/datastore.h"
+#include "Firestore/core/src/firebase/firestore/remote/stream.h"
+#include "Firestore/core/src/firebase/firestore/remote/stream_objc_bridge.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+#include "absl/strings/string_view.h"
+#include "grpcpp/support/byte_buffer.h"
+
+#import "Firestore/Source/Core/FSTTypes.h"
+#import "Firestore/Source/Model/FSTMutation.h"
+#import "Firestore/Source/Remote/FSTSerializerBeta.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+/**
+ * A Stream that implements the Write RPC.
+ *
+ * The Write RPC requires the caller to maintain special stream token
+ * state in-between calls, to help the server understand which responses the
+ * client has processed by the time the next request is made. Every response
+ * will contain a stream token; this value must be passed to the next
+ * request.
+ *
+ * After calling `Start` on this stream, the next request must be a handshake,
+ * containing whatever stream token is on hand. Once a response to this
+ * request is received, all pending mutations may be submitted. When
+ * submitting multiple batches of mutations at the same time, it's
+ * okay to use the same stream token for the calls to `WriteMutations`.
+ */
+class WriteStream : public Stream {
+ public:
+  WriteStream(util::AsyncQueue* async_queue,
+              auth::CredentialsProvider* credentials_provider,
+              FSTSerializerBeta* serializer,
+              Datastore* datastore,
+              id<FSTWriteStreamDelegate> delegate);
+
+  void SetLastStreamToken(NSData* token);
+  /**
+   * The last received stream token from the server, used to acknowledge which
+   * responses the client has processed. Stream tokens are opaque checkpoint
+   * markers whose only real value is their inclusion in the next request.
+   *
+   * `WriteStream` manages propagating this value from responses to the
+   * next request.
+   */
+  NSData* GetLastStreamToken() const;
+
+  /**
+   * Tracks whether or not a handshake has been successfully exchanged and
+   * the stream is ready to accept mutations.
+   */
+  bool IsHandshakeComplete() const {
```

Comment:
Peeking at the implementation, this looks like an accessor, so this could be named `is_handshake_complete()`. 

But you could also argue that this isn't an accessor, and just happens to be implemented with a private variable, in which case, it's already fine.

Optional.

Reply:
Done.